plat/hikey/plat_pm.c

```diff
@@ -207,10 +207,24 @@ static void hikey_affinst_suspend_finish(uint32_t afflvl,
 
 static void __dead2 hikey_system_off(void)
 {
-	gpio_set_value(0, 0);
-	mdelay(1000);
-	/* Send the system reset request since it fails to power off */
-	mmio_write_32(AO_SC_SYS_STAT0, 0x48698284);
+	NOTICE("%s: off system\n", __func__);
+
+	/* pulling GPIO_0_0 low to trigger PMIC shutdown */
+	/* setting pinmux */
+	mmio_write_32(0xF8001810, 0x2);
+	/* setting pin direction */
+	mmio_write_8(0xF8011400, 1);
+	/* setting pin output value */
+	mmio_write_8(0xF8011004, 0);
+
+	/* PMIC shutdown depends on two conditions: GPIO_0_0 (PWR_HOLD) low,
+	 * and VBUS_DET < 3.6V. For HiKey, VBUS_DET is connected to VDD_4V2
+	 * through Jumper 1-2. So, to complete shutdown, user needs to manually
+	 * remove Jumper 1-2.
+	 */
+	NOTICE(".........................................................\n");
+	NOTICE(" IMPORTANT: Remove Jumper 1-2 to complete shutdown\n");
+	NOTICE(" DANGER:    SoC is still burning. DANGER!\n");
```

Comment:
On Tue, Jun 14, 2016 at 7:30 PM, Guodong Xu <notifications@github.com> wrote:
> Reboot also has its problem. The user wants to shutdown, but he finds it
> always 'fails'. BTW, we also need a balance here between effort and result.
> If I have enough time to work on this, I will agree with you. It's better to
> check the Jumper status, and do something like blinking all 4 LEDs to catch
> user's attention, then if still fails, I can do a reset, and in the reset,
> only start one core with minimum kernel. Or, before the kernel falls into
> PSCI system-off, it can check the Jumper status. If it's connected, kernel
> should never shutdown...

I think if J1-2 is set, the board should reset on power-off. Since
J1-2 enables power-on any time the power is plugged in, it makes much
more sense having it reset then having the board hang and potentially
overheat. Even with a warning, folks may not see the serial line.

I think the best way of thinking about it is that  J1-2 means turn the
machine on any time power is applied.  So it "turns off", and then
back on!